src/v/iceberg/table_update_applier.cc

```diff
@@ -0,0 +1,172 @@
+            // -1 indicates that we should set the schema to the latest one.
+            if (meta.schemas.empty()) {
+                vlog(log.error, "Can't set -1 when there are no schemas");
+                return outcome::unexpected_state;
+            }
+            auto max_id = meta.schemas.front().schema_id;
+            for (const auto& s : meta.schemas) {
+                max_id = std::max(max_id, s.schema_id);
+            }
+            meta.current_schema_id = max_id;
+            return outcome::success;
+        }
+        auto s = std::find_if(
+          meta.schemas.begin(),
+          meta.schemas.end(),
+          [sid](const schema& schema) { return schema.schema_id == sid; });
+        if (s == meta.schemas.end()) {
+            vlog(log.error, "Schema {} doesn't exist", sid);
+            return outcome::unexpected_state;
+        }
+        meta.current_schema_id = update.schema_id;
+        return outcome::success;
+    }
+    outcome operator()(const add_spec& update) {
+        auto sid = update.spec.spec_id;
+        auto s = std::find_if(
+          meta.partition_specs.begin(),
+          meta.partition_specs.end(),
+          [sid](const partition_spec& spec) { return spec.spec_id == sid; });
+        if (s != meta.partition_specs.end()) {
+            vlog(log.error, "Partition spec id {} doesn't exist", sid);
+            return outcome::unexpected_state;
+        }
+        meta.partition_specs.emplace_back(update.spec.copy());
+        return outcome::success;
+    }
+    outcome operator()(const add_snapshot& update) {
+        auto sid = update.snapshot.id;
+        if (!meta.snapshots.has_value()) {
+            meta.snapshots.emplace();
+        }
+        auto s = std::find_if(
+          meta.snapshots->begin(),
+          meta.snapshots->end(),
+          [sid](const snapshot& snap) { return snap.id == sid; });
+        if (s != meta.snapshots->end()) {
+            vlog(log.error, "Snapshot id {} already exists", sid);
+            return outcome::unexpected_state;
+        }
+        if (update.snapshot.sequence_number <= meta.last_sequence_number) {
+            vlog(
+              log.error,
+              "New sequence number must be higher than last: {} <= {}",
+              update.snapshot.sequence_number,
+              meta.last_sequence_number);
+            return outcome::unexpected_state;
+        }
+
+        meta.snapshots->emplace_back(update.snapshot);
+        meta.last_sequence_number = update.snapshot.sequence_number;
+        return outcome::success;
+    }
+    outcome operator()(const remove_snapshots& update) {
+        if (!meta.snapshots.has_value()) {
+            return outcome::success;
+        }
+        chunked_hash_set<snapshot_id> to_remove;
+        for (const auto& id : update.snapshot_ids) {
+            to_remove.emplace(id);
+        }
+        chunked_vector<snapshot> new_list;
+        new_list.reserve(meta.snapshots->size() - to_remove.size());
```

Comment:
if the caller didn't necessarily know if meta.snapshots.has_value() was true, should it know how big meta.snapshots->size() is? i'm concerned this could go negative..

Reply:
Oh good call. Will fix